### DISCUSSION TITLE
Intercept and manipulate proxy request

### DISCUSSION THREAD

**Discussion by sureshkumarramu** [dotnet/yarp#447] [General] [answered]
Is it possible to intercept the “Request’s Response” and fill HTML forms and auto submit for specific URL !!

If this is possible, high level guidance/approach will be a great help for me.

Thanks in advance.

**Comment by Tratcher** [comment, marked as answer]
Response body manipulation is something we do not plan first class support for, but it is possible through middleware.

The ResponseCompressionMiddleware is the closest example I have for conditionally intercepting and altering the response. Note it avoids buffering the response to limit latency, but it might be simpler for you to buffer the response for specific urls, parse, edit, and resend.
https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/ResponseCompression/src/ResponseCompressionMiddleware.cs

**Reply by sureshkumarramu** [reply]
@Tratcher Thanks for the response. I will explore on the option you suggested. 

If worked, I will update in this post.

**Reply by sureshkumarramu** [reply]
Hi @Tratcher 

When I intercept and read the response body stream, I am only getting encoded string. Not sure this is related to reverse proxy, Can you please suggest how I can read response body stream as plain string.

Thanks

Attached the screenshot of the code 

![image](https://user-images.githubusercontent.com/3427928/96211103-0488e400-0fa6-11eb-867a-1e1ba1519bb3.png)

**Reply by Tratcher** [reply]
Can you share the response headers as well? They should say what the format is. Is the content compressed?

**Reply by sureshkumarramu** [reply]
I missed to notice that; it is Gzip compressed/encoded. I will explore how to decode it. Thanks @Tratcher 

![image](https://user-images.githubusercontent.com/3427928/96252696-f1e1cf80-0fe4-11eb-85dd-a64212dd0dc3.png)

**Reply by Tratcher** [reply]
https://docs.microsoft.com/en-us/dotnet/api/system.io.compression.gzipstream.-ctor?view=netcore-3.1
